sync/sync_test.go

```diff
@@ -45,26 +53,76 @@ func TestSyncBlocks(t *testing.T) {
 		testDB := pebble.NewMemTest()
 		bc := blockchain.New(testDB, utils.MAINNET)
 		synchronizer := New(bc, gw, log)
-		ctx, _ := context.WithTimeout(context.Background(), time.Second)
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 
 		require.NoError(t, synchronizer.Run(ctx))
+		cancel()
 
 		testBlockchain(t, bc)
 	})
+
+	b0, err := gw.BlockByNumber(context.Background(), 0)
+	require.NoError(t, err)
+	s0, err := gw.StateUpdate(context.Background(), 0)
+	require.NoError(t, err)
+
 	t.Run("sync multiple blocks in a non-empty db", func(t *testing.T) {
 		testDB := pebble.NewMemTest()
 		bc := blockchain.New(testDB, utils.MAINNET)
-		b0, err := gw.BlockByNumber(context.Background(), 0)
-		require.NoError(t, err)
-		s0, err := gw.StateUpdate(context.Background(), 0)
-		require.NoError(t, err)
 		require.NoError(t, bc.Store(b0, s0, nil))
-
 		synchronizer := New(bc, gw, log)
-		ctx, _ := context.WithTimeout(context.Background(), time.Second)
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 
 		require.NoError(t, synchronizer.Run(ctx))
+		cancel()
 
 		testBlockchain(t, bc)
 	})
+
+	t.Run("sync a block in an empty db, with an unreliable gw", func(t *testing.T) {
```

Comment:
What purpose would that serve? Retry mechanisms are exactly the same for empty/non-empty blockchains.

Reply:
To decouple tests from its implementation. We want to make as few assumptions about the code under test as possible. Tests are meant to describe the behaviour of the API exposed by sync package. By not writing the non-empty db test case you are assuming the retry mechanism will never be different for empty vs non-empty blockcahin. This is not necessarily true.

If we decided to expose Retry mechanism through a `Retry` func and wrote tests for the retry func then you would be correct.

Also, these are our integration tests. We should try to enumerate as many cases a possible since it is already hard to write integration tests.

Reply:
> IMO that is a very reasonable assumption to make

I see your point

> If I change the test to sync multiple blocks, would that be fine?

Yes, please.

Reply:
Done